### PR TITLE
[feature]: add daemon flag to set no_new_priv as default for unprivileged containers

### PR DESCRIPTION
**Pull request by jmzwcn**
[feature]: add daemon flag to set no_new_priv as default for unprivileged containers #29951

Signed-off-by: Daniel Zhang <jmzwcn@gmail.com>

### CODE REVIEW

**Comment by justincormack** [comment]
As commented previously, we need to also support the old syntax without `=true` if possible, and have tests for both. A test for disabling on the daemon (and another for re-enabling) is also needed.

**Comment by jmzwcn** [comment]
Actually old usage is still supported. I don't remove [here](https://github.com/jmzwcn/docker/blob/2cdffaa41d67c41aa618e2c4efecea8d0130b4c1/daemon/daemon_unix.go#L184).

**Comment by justincormack** [comment]
Ok, if you could add the tests then with the old and new forms, and a test of the daemon flag that would be great

**Comment by jmzwcn** [comment]
Sure, have done, please let me know if there is any problem.

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -159,6 +159,21 @@ func getBlkioWeightDevices(config containertypes.Resources) ([]specs.WeightDevic
 	return blkioWeightDevices, nil
 }
 
+func parseAndSetSecurityOpt(daemon *Daemon, container *container.Container, config *containertypes.HostConfig) error {
+	var findNNP bool
+	err := parseSecurityOpt(container, config)
+	for _, opt := range config.SecurityOpt {
+		if strings.Contains(opt, "no-new-privileges") {
```
Comment (justincormack) [inline daemon/daemon_unix.go]: this match is not very precise

Comment (jmzwcn) [inline daemon/daemon_unix.go]: opt case: no-new-privileges|no-new-privileges=true|no-new-privileges=false
wrong input? no-new-privileges=abc?

Comment (justincormack) [inline daemon/daemon_unix.go]: should not match : `really-no-new-privileges`, `no-new-privileges-please=true` etc

**Comment by justincormack** [comment]
The test failure is
```
12:17:31 # github.com/docker/docker/daemon
12:17:31 daemon/daemon_unix_test.go:14: imported and not used: "github.com/docker/docker/volume/drivers" as volumedrivers
12:17:31 daemon/daemon_unix_test.go:283: undefined: drivers in drivers.volumedrivers
```

which needs fixing (maybe changed at the time you rebased?)

**Comment by jmzwcn** [comment]
Have updated, please review again.

File: daemon/config_unix.go [hunk 2]
```diff
@@ -89,6 +90,7 @@ func (config *Config) InstallFlags(flags *pflag.FlagSet) {
 	flags.Int64Var(&config.CPURealtimePeriod, "cpu-rt-period", 0, "Limit the CPU real-time period in microseconds")
 	flags.Int64Var(&config.CPURealtimeRuntime, "cpu-rt-runtime", 0, "Limit the CPU real-time runtime in microseconds")
 	flags.StringVar(&config.SeccompProfile, "seccomp-profile", "", "Path to seccomp profile")
+	flags.BoolVar(&config.NoNewPrivileges, "no-new-privileges", false, "Disable container processes from gaining new privileges")
```
Comment (clnperez) [inline daemon/config_unix.go]: The way I read the linked issue this would be true by default. Is that not the case?

Comment (jmzwcn) [inline daemon/config_unix.go]: which link? I think It is false by default.

Comment (jmzwcn) [inline daemon/config_unix.go]: I think it false by default. @justincormack , what do you think?

Comment (justincormack) [inline daemon/config_unix.go]: No, false by default, to be compatible with the current state, otherwise things will break.

Comment (clnperez) [inline daemon/config_unix.go]: Thanks for the clarification @justincormack.

**Comment by clnperez** [comment]
This one: https://github.com/docker/docker/issues/29951

Cleared up now, thanks.

**Comment by jmzwcn** [comment]
Any help about this conflict(integration-cli/docker_cli_run_unix_test.go)?

**Comment by jmzwcn** [comment]
Any review?

**Comment by justincormack** [comment]
Sorry, its a bit hectic with getting 1.13 released, I will take a look
again soon...

**Comment by jmzwcn** [comment]
Could we please have a review ASAP? I am afraid there will be conflict again.

File: docs/reference/commandline/dockerd.md [hunk 3]
```diff
@@ -68,7 +68,8 @@ Options:
       --max-concurrent-downloads int          Set the max concurrent downloads for each pull (default 3)
       --max-concurrent-uploads int            Set the max concurrent uploads for each push (default 5)
       --metrics-addr string                   Set address and port to serve the metrics api (default "")
-      --mtu int                               Set the containers network MTU
+      --mtu int                               Set the containers network mtu
```
Comment (justincormack) [inline docs/reference/commandline/dockerd.md]: Don't make this unrelated change (MTU is normal anyway)

File: docs/reference/commandline/dockerd.md [hunk 4]
```diff
@@ -68,7 +68,8 @@ Options:
...
+      --no-new-privileges                     Disable container processes from gaining new privileges
```
Comment (justincormack) [inline docs/reference/commandline/dockerd.md]: lower in this file there is also a section "This is a full example of the allowed configuration options on Linux:" which should also have this added as `"no-new-privileges":false`

**Comment by justincormack** [comment]
Two minor comments but otherwise this looks fine, tested it thoroughly and it is working fine.  Thanks very much for doing this.

File: docs/reference/commandline/dockerd.md (hunk 4 above)
Comment (jmzwcn) [inline docs/reference/commandline/dockerd.md]: Seems not get what you said:
--no-new-privileges                     Disable container processes from gaining new privileges
how to modify?

Comment (justincormack) [inline docs/reference/commandline/dockerd.md]: sorry I just mean in this section https://github.com/docker/docker/blob/master/docs/reference/commandline/dockerd.md#linux-configuration-file it needs to be added in the example config file.

**Comment by jmzwcn** [comment]
Any question?

**Review by justincormack — APPROVED** [review]

**Comment by justincormack** [comment]
LGTM cc @jessfraz @runcom @mrunalp

File: daemon/daemon_unix.go [hunk 5]
```diff
@@ -159,6 +159,21 @@ func getBlkioWeightDevices(config containertypes.Resources) ([]specs.WeightDevic
 	return blkioWeightDevices, nil
 }
 
+func parseAndSetSecurityOpt(daemon *Daemon, container *container.Container, config *containertypes.HostConfig) error {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: What's the purpose of this new function?
Isn't `parseSecurityOpt` already setting some values?

File: daemon/config_unix.go (hunk 2 above)
Comment (cpuguy83) [inline daemon/config_unix.go]: Maybe `default-no-new-privileges` otherwise it looks like it's doing something to do the daemon.

File: daemon/daemon_unix.go (hunk 5 above)
Comment (jmzwcn) [inline daemon/daemon_unix.go]: Set NNP value, we couldn't get daemon in old "parseSecurityOpt" func.

Comment (cpuguy83) [inline daemon/daemon_unix.go]: We don't really need daemon, we only need to pass in the default value.

Comment (jmzwcn) [inline daemon/daemon_unix.go]: If no daemon is passed, we need "daemon.configStore.NoNewPrivileges".

Comment (cpuguy83) [inline daemon/daemon_unix.go]: Indeed, or something like a `securityOpt` type.

Comment (jmzwcn) [inline daemon/daemon_unix.go]: Seems it is a bit hard to pass daemon.configStore.NoNewPrivileges, because it only exist in unix, but our parseAndSetSecurityOpt is across platform...

**Comment by jmzwcn** [comment]
@cpuguy83 
Seems it is a bit hard to pass daemon.configStore.NoNewPrivileges, because it only exist in unix, but our parseAndSetSecurityOpt is across platform...

File: daemon/daemon_unix.go (hunk 5 above)
Comment (cpuguy83) [inline daemon/daemon_unix.go]: There's a few things that could be done here.
First that comes to mind is to pass around a function that gets default security options.

**Comment by jmzwcn** [comment]
I have restored the previous version. Please let me know if there is any problem.

File: daemon/daemon_unix.go (hunk 5 above)
Comment (jmzwcn) [inline daemon/daemon_unix.go]: you means before pass, we set this value from daemon in,such as createContainerPlatformSpecificSettings?

Comment (cpuguy83) [inline daemon/daemon_unix.go]: My main point here is that passing daemon in as an argument defeats the purpose of having this as a stand-alone function.
This might as well be
```go
func(d *Daemon) parseSecurityOpt(...) {
    // do stuff
    parseSecurityOpt(...)
}

**Comment by LK4D4** [comment]
ping @cpuguy83 were your comments addressed?

**Comment by jessfraz** [comment]
omg my dreams are coming trueeeeeee

File: daemon/daemon_unix.go [hunk 6]
```diff
@@ -159,6 +159,21 @@ func getBlkioWeightDevices(config containertypes.Resources) ([]specs.WeightDevic
...
+	var findNNP bool
+	err := parseSecurityOpt(container, config)
```
Comment (jessfraz) [inline daemon/daemon_unix.go]: it seems like youd want to return the error early here instead of iterating thru all the opts then returning the error? or maybe add a comment as to why this is ok

**Comment by jessfraz** [comment]
might be nice to also add an integration test for the daemon where `no_new_privs` is true but a container is run with `privileged` and then one where you override `no_new_privs` to false on the container. :innocent:

File: daemon/daemon_unix.go (hunk 5 above)
Comment (jessfraz) [inline daemon/daemon_unix.go]: if this is going to apply to only the daemon, maybe it should be `parseAndSetDaemonSecurityOpt`

Comment (jessfraz) [inline daemon/daemon_unix.go]: or i like more what @cpuguy said above

**Comment by jmzwcn** [comment]
@cpuguy83 ,
func(d *Daemon) parseSecurityOpt(...) { **//Actually this method exist in container.go, because this setting is specified to daemon_unix.go**
    // do stuff--**will have to make OS condition, it be best-practice?**
    parseSecurityOpt(...)
}

**Comment by cpuguy83** [comment]
@jmzwcn It can go in a platform file, just that it will need to be defined on each platform.

**Comment by jmzwcn** [comment]
@cpuguy83, already update, please let me know if any problem.

@jessfraz , seems TestParseNNPSecurityOptions has covered that scenario.

**Review by cpuguy83 — APPROVED** [review]
Nice, looks much cleaner. Thanks!

LGTM

**Comment by jmzwcn** [comment]
The conflict happen again. If possible, please merge ASAP. Thanks!

**Comment by cpuguy83** [comment]
ping @justincormack

File: integration-cli/docker_cli_run_unix_test.go [hunk 7]
```diff
@@ -1140,13 +1140,26 @@ func (s *DockerSuite) TestRunSeccompDefaultProfileNS(c *check.C) {
...
 
 	// test that running a setuid binary results in no effective uid transition
+	icmd.RunCommand(dockerBinary, "run", "--security-opt", "no-new-privileges=true", "--user", "1000",
+		"nnp-test", "/usr/bin/nnp-test").Assert(c, icmd.Expected{
+		Out: "EUID=1000",
+	})
+}
+
+// TestLegacyRunNoNewPrivSetuid checks that --security-opt=no-new-privileges prevents
+// effective uid transtions on executing setuid binaries.
```
Comment (php-coder) [inline integration-cli/docker_cli_run_unix_test.go]: s/transtions/transitions/

**Comment by thaJeztah** [comment]
The actual _flag_ to set this got lost in a rebase, and was added again in https://github.com/moby/moby/pull/32944
